File: python/excercise84/largest_rectangle_in_histogram.py
from typing import List


def largest_rectangle_in_histogram(heights: List[int]) -> int:
    stk = [-1]
    max_area = 0
    for i in range(0, len(heights) + 1):
        curr_height = 0 if i == len(heights) else heights[i]
        while len(stk) > 1 and curr_height < heights[stk[-1]]:
            h = heights[stk.pop()]
            area = (i - 1 - stk[-1]) * h
            max_area = max(max_area, area)
        stk.append(i)
    return max_area

File: go/excercise84/largest_rectangle_in_histogram.go
package largest_rectangle_in_histogram

func largestRectangleArea(heights []int) int {
	stack := make([]int, 0)
	stack = append(stack, -1) // 第一个柱子的下标是0，默认他前面一个是-1。

	maxArea := 0

	for i := 0; i <= len(heights); i++ {
		var curHeight int
		if i < len(heights) {
			curHeight = heights[i]
		} else {
			curHeight = 0
		}

		for len(stack) > 1 && curHeight < heights[stack[len(stack)-1]] {
			h := heights[stack[len(stack)-1]]
			stack = stack[:len(stack)-1]

			area := (i - 1 - stack[len(stack)-1]) * h

			maxArea = max(maxArea, area)
		}

		stack = append(stack, i)
	}

	return maxArea
}
